### ISSUE TITLE
Burn 0.21 — `Optimizer::step` is a silent no-op for `#[derive(Module)]` models

### ISSUE THREAD

**Issue by vexschwarz** [tracel-ai/burn#5014]
# Burn 0.21 — `Optimizer::step` is a silent no-op for `#[derive(Module)]` models

> Bug report draft for upstream filing at https://github.com/tracel-ai/burn/issues.
> Discovered on Burn 0.21.0, reproduced on both `Autodiff<Cuda<f32, i32>>`
> and `Autodiff<NdArray<f32>>`, with both `AdamWConfig` and `SgdConfig`.

---

## TL;DR

`Optimizer::step` returns a model whose `Param` values are **byte-identical**
to the input model, despite:

- `loss.backward()` producing finite gradients,
- `param.grad(&grads)` resolving to the expected non-zero tensors directly,
- the per-Param `ParamId` being preserved across the step.

The breakage is at the `GradientsParams::from_grads` → `module.visit` →
`GradientsParamsConverter::visit_float` chain: the converter populates an
empty `TensorContainer` (`grads_params.len() == 0`), so the downstream
`OptimizerAdaptor::step → module.map(&mut SimpleOptimizerMapper)` chain
finds no grad to apply and silently returns the unmodified Param.

A user-side workaround (manual update via the public `Param::map`) recovers
expected SGD behaviour.

---

## Minimal Reproduction

```rust
use burn::backend::{Autodiff, NdArray};
use burn::backend::ndarray::NdArrayDevice;
use burn::module::{Module, Param, ParamId};
use burn::nn::{Linear, LinearConfig};
use burn::optim::{AdamWConfig, GradientsParams, Optimizer};
use burn::tensor::backend::Backend;
use burn::tensor::{Distribution, Tensor};

type B = Autodiff<NdArray<f32>>;

#[derive(Module, Debug, Clone)]
struct TinyModel<BK: Backend> {
    emb: Param<Tensor<BK, 2>>,
    head: Linear<BK>,
}

impl<BK: Backend> TinyModel<BK> {
    fn init(device: &BK::Device) -> Self {
        // Canonical Param::uninitialized — same pattern Linear/Embedding use.
        let emb = Param::uninitialized(
            ParamId::new(),
            |device, require_grad| {
                let t = Tensor::<BK, 2>::random(
                    [4, 8],
                    Distribution::Normal(0.0, 0.1),
                    device,
                );
                if require_grad { t.require_grad() } else { t }
            },
            device.clone(),
            true,
            [4, 8].into(),
        );
        Self {
            emb,
            head: LinearConfig::new(8, 1).init(device),
        }
    }
}

fn snapshot<BK: Backend, const D: usize>(p: &Param<Tensor<BK, D>>) -> Vec<f32> {
    p.val().into_data().as_slice::<f32>().unwrap().to_vec()
}

fn main() {
    let device = NdArrayDevice::default();
    let mut model: TinyModel<B> = TinyModel::init(&device);
    let mut opt = AdamWConfig::new().init::<B, TinyModel<B>>();

    let pre_emb = snapshot(&model.emb);
    let pre_w = snapshot(&model.head.weight);

    // Forward + backward with a synthetic loss that depends on `emb` AND `head`.
    let x: Tensor<B, 2> = Tensor::random([4, 8], Distribution::Default, &device);
    let y: Tensor<B, 2> = Tensor::random([4, 1], Distribution::Default, &device);
    let emb_mean: Tensor<B, 2> = model.emb.val().mean_dim(0);
    let logits = model.head.forward(x + emb_mean);
    let loss = (logits - y).powf_scalar(2.0).mean();

    let grads = loss.backward();

    // Direct per-Param grad lookup works.
    let emb_grad_direct = model.emb.grad(&grads).expect("emb grad");
    let w_grad_direct   = model.head.weight.grad(&grads).expect("weight grad");
    println!(
        "direct emb_grad max-abs = {}",
        emb_grad_direct.into_data().as_slice::<f32>().unwrap()
            .iter().map(|v| v.abs()).fold(0.0_f32, f32::max),
    );
    println!(
        "direct weight_grad max-abs = {}",
        w_grad_direct.into_data().as_slice::<f32>().unwrap()
            .iter().map(|v| v.abs()).fold(0.0_f32, f32::max),
    );

    // But `GradientsParams::from_grads` returns an empty container.
    let grads_params = GradientsParams::from_grads::<B, TinyModel<B>>(grads, &model);
    println!("grads_params.len() = {}", grads_params.len()); // expected: 3, actual: 0

    // Optimizer step is a silent no-op as a consequence.
    model = opt.step(1e-1, model, grads_params);

    let post_emb = snapshot(&model.emb);
    let post_w   = snapshot(&model.head.weight);

    let emb_delta: f32 = pre_emb.iter().zip(&post_emb)
        .map(|(a, b)| (a - b).powi(2)).sum::<f32>().sqrt();
    let w_delta: f32 = pre_w.iter().zip(&post_w)
        .map(|(a, b)| (a - b).powi(2)).sum::<f32>().sqrt();
    println!("‖Δ emb‖    = {emb_delta}");   // observed: 0
    println!("‖Δ weight‖ = {w_delta}");     // observed: 0
}
```

### Observed output (Burn 0.21.0)

```
direct emb_grad max-abs    = 0.023...
direct weight_grad max-abs = 0.587...
grads_params.len() = 0
‖Δ emb‖    = 0
‖Δ weight‖ = 0
```

### Expected output

```
grads_params.len() = 3                        (emb, head.weight, head.bias)
‖Δ emb‖    > 0
‖Δ weight‖ > 0
```

---

## Diagnosis Matrix

I ran an isolation matrix to eliminate every user-side variable; all eight
experiments observed `‖Δ‖ = 0` on the optimizer pipeline:

| # | Variable swapped | Result |
|---|---|---|
| 1 | Backend: `Autodiff<Cuda>` → `Autodiff<NdArray<f32>>` | Δ = 0 (no-op survives) |
| 2 | Optimizer: `AdamWConfig` → `SgdConfig` (Cuda kept) | Δ = 0 |
| 3 | Removed nested submodule wrapper — flat struct only (`emb`, `head`) | Δ = 0 |
| 4 | Param init: `Param::from_tensor` → `Param::uninitialized` lazy closure | Δ = 0 |
| 5 | `Linear`-only `Head` (canonical `LinearConfig::init` Param) + AdamW | Δ = 0 |
| 6 | Instrumented `GradientsParams::from_grads` — inspected `.len()` | `len() == 0` (root cause area) |
| 7 | Manual `grads_params.register(param.id, grad)` followed by `opt.step` | Δ = 0 (`opt.step` still no-ops) |
| 8 | Bypass `opt.step` entirely — manual `Param::map(\|t\| t - lr * grad)` | **Δ > 0 ✅** |

So the breakage is somewhere in the chain:

```
loss.backward()  →  GradientsParams::from_grads  →  OptimizerAdaptor::step
                       └─ module.visit             └─ module.map
                          └─ visit_float             └─ map_float
                             └─ param.val()            └─ grads.remove(id)
                                .grad_remove(grads)
```

Manual `param.map(...)` (Burn's own public API at
`burn-core-0.21.0/src/module/param/base.rs:294`) bypasses every step in the
above chain and works correctly — strongly suggesting the bug is *not* in
`Param::map` itself but in the visitor/mapper traversal performed by
`#[derive(Module)]`'s emitted `visit`/`map` implementations or in
`GradientsParamsConverter::visit_float`'s `param.val().grad_remove(grads)`.

---

## Hypothesis — `param.val().grad_remove(grads)` returns `None`

`GradientsParamsConverter::visit_float`
(`burn-optim-0.21.0/src/optim/visitor.rs:31–44`) does:

```rust
fn visit_float<const D: usize>(&mut self, param: &Param<Tensor<B, D>>) {
    // ...
    let Some(grad) = param.val().grad_remove(self.grads) else {
        return;  // ← every Param hits this early return
    };
    self.grads_params.register::<B::InnerBackend, D>(param.id, grad);
}
```

A direct `param.grad(&grads)` call from user code (which goes through the
`Param.id` path) **does** return the expected gradient. But
`param.val().grad_remove(grads)` looks up the grad via the *tensor's*
internal `ParamId` metadata (set during `require_grad()` on the autodiff
graph).

The two `ParamId`s appear to diverge in 0.21:

- `Param.id` is generated by `ParamId::new()` at construction time
  (`Param::from_tensor`, `Param::uninitialized`, …) and is what
  `param.grad()` uses to query the gradient store.
- The tensor's autodiff-tracked `ParamId` is allocated when
  `tensor.require_grad()` is first applied; that `ParamId` is what
  `loss.backward()` registers in the `Gradients` container.

If these two `ParamId`s do not match (or if `Param::val()`'s
`SyncOnceCell::get_or_init` clone strips the autograd `ParamId` metadata
from the returned tensor), then `grad_remove(grads)` returns `None` for
every Param.

I have not yet patched Burn to add `eprintln!` confirmation, but this is
the most parsimonious explanation for:

1. `param.grad(&grads)` succeeding directly (`Param.id` keyed lookup).
2. `grads_params.len() == 0` after `from_grads` (visitor's `val()`-derived
   `ParamId` keyed lookup misses).
3. `OptimizerAdaptor::step → module.map → SimpleOptimizerMapper::map_float`
   then calling `self.grads.remove(id)` with an empty container and
   returning the tensor unchanged (`simple/adaptor.rs:182,234`).
4. Manual `Param::map(|t| t - lr * grad)` working — it never touches the
   visitor/mapper chain.

---

## Workaround Used Downstream

```rust
use burn::module::Param;
use burn::tensor::{Tensor, backend::AutodiffBackend};

pub fn manual_sgd_step_2d<B: AutodiffBackend>(
    param: Param<Tensor<B, 2>>,
    grad: Tensor<B::InnerBackend, 2>,
    lr: f32,
) -> Param<Tensor<B, 2>> {
    param.map(|t| {
        let is_require_grad = t.is_require_grad();
        let inner = t.inner();
        let updated = inner - grad.mul_scalar(lr);
        let mut new_t = Tensor::<B, 2>::from_inner(updated);
        if is_require_grad { new_t = new_t.require_grad(); }
        new_t
    })
}
```

Call site:

```rust
let grads = loss.backward();
let emb_grad = model.emb.grad(&grads).unwrap();
model.emb = manual_sgd_step_2d(model.emb, emb_grad, lr);
```

This restores the expected `‖Δ‖ > 0` and a falling loss across epochs.

---

## Environment

- **Burn**: 0.21.0 (registry crates `burn`, `burn-core`, `burn-cuda`,
  `burn-ndarray`, `burn-optim`, `burn-derive`, all 0.21.0).
- **Cargo dependencies**:
  ```toml
  burn      = { version = "0.21", default-features = false, features = ["std", "ndarray", "autodiff", "train"] }
  burn-cuda = { version = "0.21", default-features = false }
  cubecl    = { version = "0.10", default-features = false, features = ["std", "cuda"] }
  ```
- **GPU**: RTX 5090 (sm_120) — reproduced.
- **CPU NdArray backend** — also reproduced (rules out CUDA driver / TMA /
  cubecl-cuda).
- **rustc**: 1.94.1
- **OS**: Ubuntu 25.04, Linux 6.14

---

## Diagnosis binaries (available on request)

Eight standalone diagnosis binaries reproduce each row of the matrix above
in isolation. The three most decisive ones for upstream investigators:

- `diag_a6_grads_inspect` — proves `grads_params.len() == 0` despite
  finite per-Param grads.
- `diag_a7_manual_grads` — proves manually-built `grads_params` (skipping
  `from_grads` entirely) does **not** rescue `opt.step` either — the
  mapper still does not surface deltas.
- `diag_a8_manual_param_map` — proves `Param::map` direct update works
  on the same model, on Cuda, with the same gradients. **This is the
  cleanest reproduction of the contrast.**

Happy to share sources separately if useful.

---

## Suspected files to inspect upstream

1. `burn-optim-0.21.0/src/optim/visitor.rs:31–44` —
   `GradientsParamsConverter::visit_float`. Is `param.val().grad_remove(self.grads)`
   genuinely consulting the same `ParamId` that
   `loss.backward()` populated in `self.grads`?
2. `burn-core-0.21.0/src/module/param/base.rs:256–269` — `Param::val()`.
   Does `SyncOnceCell::get_or_init(...).clone()` preserve the autodiff
   `ParamId` metadata on the returned tensor?
3. `burn-core-0.21.0/src/module/param/tensor.rs:69–72` — `Param::from_tensor`.
   Does `value.require_grad()` allocate a fresh autograd `ParamId` independent
   of the `ParamId::new()` stored on `Param.id`? If so, the visitor’s
   `grad_remove` will look at the autograd `ParamId` while
   `register::<_, D>(param.id, grad)` writes under `Param.id` — but in this
   reproduction `register` never runs (early return).
4. `burn-optim-0.21.0/src/optim/simple/adaptor.rs:178–238` —
   `SimpleOptimizerMapper::map_float`. Confirmed via diag_a7 that
   `self.grads.remove(id)` returns `None` even when the container had been
   manually populated with the correct `Param.id`; this points to
   `OptimizerAdaptor::step_common`'s `module.map(...)` traversal failing
   to visit Param fields generated by `#[derive(Module)]`.

---

## Suggested fix direction

Two non-exclusive candidates:

- **A.** Have `Param::val()` re-attach the autograd `ParamId` metadata equal
  to `self.id` on every returned clone, so that the visitor's
  `grad_remove(grads)` lookup matches the same key as `param.grad()`.
- **B.** Change `GradientsParamsConverter::visit_float` to lookup via
  `param.grad(self.grads)` (which uses `Param.id`) instead of
  `param.val().grad_remove(self.grads)` (which uses the tensor-side
  `ParamId`). This aligns the optimizer pipeline with the public
  `Param::grad` API.

(B) is the smaller patch and matches the API surface that already works
correctly from user code.

---

## Acknowledgements

Burn is a great Rust-native ML stack — this report is filed in good faith so
that the optimizer pipeline can match the high quality of the rest of the
framework. Happy to provide additional traces, GPU dumps, or reproduce on any
specific commit.

**Comment by laggui** [comment]
The result seems a bit unintuitive, but the `#[derive(Module)]` requires the the backend generic to always be `B`. 

In this case though it's just treated as a constant module, which means the generated module code is empty. Therefore no visitation actually happens.

https://github.com/tracel-ai/burn/blob/f3a81f0af111cc15c9b59c2d12379a23440c810c/crates/burn-derive/src/module/base.rs#L9-L26

https://github.com/tracel-ai/burn/blob/f3a81f0af111cc15c9b59c2d12379a23440c810c/crates/burn-derive/src/shared/generics.rs#L39-L62

If you simply change to `B: Backend` it will work.
```rust
#[derive(Module, Debug, Clone)]
struct TinyModel<B: Backend> {
    emb: Param<Tensor<B, 2>>,
    head: Linear<B>,
}
```

And in future releases there is no more backend generic 🙂 

Sorry for your troubles.
